### PR TITLE
Add traversable paginator

### PR DESCRIPTION
**Pull request by toby-griffiths**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tickets       | n/a
| License       | MIT
| Doc PR        | api-platform/docs#1194

I've had need of a paginator for my custom data provider that returns a Generator, so I thought I'd submit this as an addition to the core, if that's so desired?

* [x] Add tests and ensure they pass.
* [x] Never break backward compatibility.
* [x] Features and deprecations must be submitted against master branch.
* [x] Update CHANGELOG.md file.
* [x] Submit PR for docs update.

### CODE REVIEW

**Comment by toby-griffiths** [comment]
Not sure what to do about the failed build checks.  The skipped tests that seems be be causing a failure are unrelated to my change, and now I'm seeing a `Cannot update only a partial set of packages without a lock file present.` error.  On other projects I could click to re-run the build checks, but I don't seem to have that option here.

**Comment by soyuka** [comment]
we fixed our tests feel free to rebase to relaunch the CI

**Review by soyuka — APPROVED** [review]

**Comment by toby-griffiths** [comment]
@soyuka I've rebased & the build is still failing on SwaggerUiAction for both phpcs & phpunit checks.  I can take a look at these in a fresh PR if you like?

There also looks to be a Github authentication issue?

**Comment by soyuka** [comment]
Yes there's an issue with github authentification using composer 2 I haven't got the time to dig into that more :/

**Review by soyuka — APPROVED** [review]

**Comment by toby-griffiths** [comment]
I know that feeling.  If I can find a little time I'll see if I can help out.

Would you be OK to add the `hacktoberfest-accepted` label in the meantime, as your approval doesn't seem to have registered on my Hacktober profile?

**Comment by toby-griffiths** [comment]
Thank you @soyuka

File: src/DataProvider/TraversablePaginator.php [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\DataProvider;
+
+use Traversable;
+
+class TraversablePaginator implements \IteratorAggregate, PaginatorInterface
```
Comment (dunglas) [inline src/DataProvider/TraversablePaginator.php:18]:
```suggestion
final class TraversablePaginator implements \IteratorAggregate, PaginatorInterface
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -19,6 +19,7 @@
 * IriConverter: Fix IRI url double encoding - may cause breaking change as some characters no longer encoded in output (#3552)
 * OpenAPI: **BC** Replace all characters other than `[a-zA-Z0-9\.\-_]` to `.` in definition names to be compliant with OpenAPI 3.0 (#3669)
 * Add stateless ApiResource attribute
+* Add TraversablePaginator
```
Comment (dunglas) [inline CHANGELOG.md:22]: Can you add it to a new `2.7.0` section please? We'll merge this new feature in the next version.

File: src/DataProvider/TraversablePaginator.php [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        return $this->currentPage;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getLastPage(): float
+    {
+        if (0. === $this->itemsPerPage) {
+            return 1;
```
Comment (dunglas) [inline src/DataProvider/TraversablePaginator.php:47]:
```suggestion
            return 1.;
```

File: src/DataProvider/TraversablePaginator.php [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+use Traversable;
+
+class TraversablePaginator implements \IteratorAggregate, PaginatorInterface
+{
+    private $taverable;
+    private $currentPage;
+    private $itemsPerPage;
+    private $totalItems;
+
+    public function __construct(Traversable $iterator, float $currentPage, float $itemsPerPage, float $totalItems)
```
Comment (dunglas) [inline src/DataProvider/TraversablePaginator.php:25]:
```suggestion
    public function __construct(\Traversable $iterator, float $currentPage, float $itemsPerPage, float $totalItems)
```

File: src/DataProvider/TraversablePaginator.php [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\DataProvider;
+
+use Traversable;
```
Comment (dunglas) [inline src/DataProvider/TraversablePaginator.php:16]:
```suggestion
```

**Review by dunglas — CHANGES_REQUESTED** [review]

**Comment by alanpoulain** [comment]
Thank you @toby-griffiths!

**Pull request merged**